cmake/Stdlib.cmake

```diff
@@ -1,33 +1,229 @@
+    write_stdlib_bitcode_lib(${name} ${target} ${os} ${bit})
+
+    add_custom_command(
+        OUTPUT ${bc}
+        COMMAND ${ispc_name} --nostdlib --gen-stdlib --target=${target} --arch=${arch} --target-os=${canon_os} stdlib.ispc --emit-llvm -o ${bc}
+        DEPENDS ${ispc_name} ${STDLIB_ISPC_FILES}
+        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+    )
+
+    string(TOUPPER ${os} OS_UP)
     add_custom_command(
-        OUTPUT ${output}
-        COMMAND ${CLANG_EXECUTABLE} -E -x c -DISPC_MASK_BITS=${mask} -DISPC=1 -DPI=3.14159265358979
-            stdlib.ispc | \"${Python3_EXECUTABLE}\" stdlib2cpp.py mask${mask}
-            > ${output}
-        DEPENDS stdlib.ispc
+        OUTPUT ${cpp}
+        COMMAND ${Python3_EXECUTABLE} bitcode2cpp.py ${bc} --type=stdlib --runtime=${bit} --os=${OS_UP} ${cpp}
+        DEPENDS bitcode2cpp.py ${bc}
         WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
     )
-    set(${outputPath} ${output} PARENT_SCOPE)
-    set_source_files_properties(${outputPath} PROPERTIES GENERATED true)
+
+    set(tmp_list_bc ${STDLIB_BC_FILES})
+    list(APPEND tmp_list_bc ${bc})
+    set(STDLIB_BC_FILES ${tmp_list_bc} PARENT_SCOPE)
+
+    set(tmp_list_cpp ${STDLIB_CPP_FILES})
+    list(APPEND tmp_list_cpp ${cpp})
+    set(STDLIB_CPP_FILES ${tmp_list_cpp} PARENT_SCOPE)
 endfunction()
 
-function(generate_stdlib resultList)
-    foreach (m ${ARGN})
-        create_stdlib(${m} outputPath)
-        list(APPEND tmpList "${outputPath}")
-        if(MSVC)
-            # Group generated files inside Visual Studio
-            source_group("Generated Stdlib" FILES ${outputPath})
-        endif()
+function (generate_stdlibs_1 ispc_name)
+    list(APPEND os_list)
+    if (ISPC_WINDOWS_TARGET)
+        list(APPEND os_list "windows")
+    endif()
+    if (ISPC_UNIX_TARGET)
+        list(APPEND os_list "unix")
+    endif()
+
+    if (NOT os_list)
+        message(FATAL_ERROR "Windows or Linux target has to be enabled")
+    endif()
+
+    # "Regular" targets, targeting specific real ISA: sse/avx
+    if (X86_ENABLED)
+        foreach (target ${X86_TARGETS})
+            foreach (bit 32 64)
+                foreach (os ${os_list})
+                    stdlib_to_cpp(${ispc_name} ${target} ${bit} ${os})
+                endforeach()
+            endforeach()
+        endforeach()
+    endif()
+
+    # XE targets
+    if (XE_ENABLED)
+        foreach (target ${XE_TARGETS})
+            # No cross-compilation for XE targets
+            if (WIN32)
+                stdlib_to_cpp(${ispc_name} ${target} 64 windows)
+            elseif (APPLE)
+                # no xe support
+            else()
+                stdlib_to_cpp(${ispc_name} ${target} 64 unix)
+            endif()
+        endforeach()
+    endif()
+
+    # ARM targets
+    if (ARM_ENABLED)
+        foreach (os ${os_list})
+            foreach (target ${ARM_TARGETS})
+                if (${os} STREQUAL "windows")
+                    continue()
+                endif()
```

Comment:
Should this condition be one level up under `foreach (os ${os_list})`?

Reply:
No, there is two more entries after the innermost loop.